sycl/include/CL/sycl/INTEL/esimd/detail/emu/esimdcpu_device_interface.hpp

```diff
@@ -0,0 +1,42 @@
+//==----- esimdcpu_device_interface.hpp - DPC++ Explicit SIMD API ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+/// \file esimdcpu_device_interface.hpp
+/// Declarations for ESIMD_CPU-device specific definitions.
+/// ESIMD intrinsic and LibCM functionalities required by intrinsic defined
+///
+/// This interface is for ESIMD intrinsic emulation implementations
+/// such as slm_access to access ESIMD_CPU specific-support therefore
+/// it has to be defined and shared as include directory
+///
+/// \ingroup sycl_pi_esimd_cpu
+
+#pragma once
+
+#include <cstdint>
+
+#ifdef _MSC_VER
+// Definitions for type consistency between ESIMD_CPU and CM_EMU
+typedef unsigned int uint;
+typedef unsigned short ushort;
+typedef unsigned char uchar;
+#endif // _MSC_VER
+
+struct ESIMDDeviceInterface {
```

Comment:
Following up our internal discussion. Even though virtual function table layout is stable and in practice compatible between clang/gcc/msvc, as @romanovvlad pointed out, C++ does not guarantee it and this potentially may cause problems if we start extending this interface. So we need to switch from this lazy way of creating the table to bulletproof one.

Something like this C-style w/o inheritance:

```cpp
// A table of function pointers implementing ESIMD emulation device functions
// which need external library support (such as libCM). We don't want to link user-app
// or SYCL runtime to these external libraries, instead, it is this plugin who links with them.
// It then fills the function pointer table and conveys it to the user app (header-side
// implementations of __esimd* intrinsics) via a specific PI extension.
// The only change allowed to this table over time is adding new member to the end.
// Whenever addition happens, the version field must be incremented so that
// it can be checked if given table supports needed functionality.
struct ESIMDDeviceInterfaceV0 {
  uintptr_t version; // interface version, initialized to 0
#include "esimd_emu_functions_v0.h"
};

// future extension would look like
struct ESIMDDeviceInterfaceV1 {
  uintptr_t version; // interface version, initialized to 1
#include "esimd_emu_functions_v0.h"
#include "esimd_emu_functions_v1.h"
};
```
`uintptr_t` for `version` to have the same size as pointers coming later to eliminate the risk of misalignment.
Possible implementation of `__esimd_feature_from_v1`:
```cpp
  void __esimd_feature_from_v1() {
    if (getESIMDDeviceInterface()->version < 1) {
      throw sycl::exception("esimd_feature_from_v1 unsupported by the plugin");
    }
    reinterpret_cast<ESIMDDeviceInterfaceV1*>(getESIMDDeviceInterface())->esimd_feature_from_v1();
  }
```

Reply:
I'll apply this approach to this PR.